File: common/chat_store.py
"""
Persistence for the main Chat page's conversations.

A chat is one of the few things on this service a person actually accumulates:
what they asked, what the agent answered, which runs it spawned, which files it
touched. It used to live in ``localStorage`` under ``agent_hub_chats_v1``, which
made it a property of one browser profile rather than of the service — gone with
the site data, absent on the next device, and silently trimmed oldest-first once
the 5 MB quota was hit. This module puts it where the rest of the state already
lives: the shared SQLite database (:mod:`common.db`), one row per conversation.

Shape
-----
The row stores the conversation **as the UI renders it**: metadata (title,
target, workspace, project) plus the ordered bubbles, each with the extras a
bubble carries (run id, token tallies, tool calls, touched files, a rich view).
That is deliberate. A transcript is reconstructible from the runs table (the
Telegram thread is rebuilt exactly that way), but only as bare user/assistant
text — the reconstruction loses which bubble a view belonged to and what the
user saw. The runs table remains the ledger of what executed; this table is the
record of the conversation that executed it.

Writes replace the whole document, the way the browser used to rewrite the whole
localStorage key. A conversation is small next to a run payload, and a
full-document write keeps the store honest about being a mirror of the client's
state rather than a second source of truth that can drift from it.

Two bounds keep one runaway chat from growing without limit: ``MAX_MESSAGES``
per chat (oldest bubbles drop first, the same order localStorage used) and
``MAX_DOC_BYTES``, after which the heavy per-bubble extras are shed before the
text is. Both are far above any hand-written conversation.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from common import db

#: How many bubbles one chat keeps. Beyond this the oldest are dropped, which is
#: what the browser store did to whole conversations once the quota was hit.
MAX_MESSAGES = 2000

#: Ceiling on one stored conversation. Bubbles carry streamed extras (tool
#: calls, file lists, view payloads) that dwarf their text, so an oversized doc
#: sheds those first and only then drops bubbles.
MAX_DOC_BYTES = 4 * 1024 * 1024

#: Per-bubble fields worth dropping before any message is lost.
_HEAVY_MSG_FIELDS = ("tool_calls", "files", "view", "timeline", "thinking_live",
                     "running_tool", "board")

#: Fields the client owns and the row mirrors into columns.
_COLUMNS = ("title", "workspace", "project_id", "agent_id", "flow_id",
            "team_id", "target_mode", "origin")


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _messages(chat: Dict[str, Any]) -> List[dict]:
    msgs = chat.get("messages")
    return [m for m in msgs if isinstance(m, dict)] if isinstance(msgs, list) else []


def _fit(chat: Dict[str, Any]) -> Dict[str, Any]:
    """Bring one conversation inside the stored-size bounds."""
    msgs = _messages(chat)
    if len(msgs) > MAX_MESSAGES:
        msgs = msgs[-MAX_MESSAGES:]
    chat = {**chat, "messages": msgs}
    if len(db.dumps(chat)) <= MAX_DOC_BYTES:
        return chat

    # Too big: shed the streamed extras first — they are display detail that the
    # run record can still answer for, while the text of the turn cannot be
    # recovered from anywhere else.
    stripped = []
    for m in msgs:
        copy = {k: v for k, v in m.items() if k not in _HEAVY_MSG_FIELDS}
        stripped.append(copy)
    chat = {**chat, "messages": stripped}
    while len(stripped) > 1 and len(db.dumps(chat)) > MAX_DOC_BYTES:
        # Drop from the front: a conversation reads from its tail.
        stripped = stripped[len(stripped) // 4 + 1:]
        chat = {**chat, "messages": stripped}
    return chat


def _summary(chat: Dict[str, Any]) -> Dict[str, Any]:
    """A chat without its transcript, for the sidebar list.

    The list renders a title and a timestamp per row, so shipping every bubble
    of every conversation to draw it would be the localStorage problem in a new
    place: a page load proportional to the whole history.
    """
    msgs = _messages(chat)
    last = msgs[-1] if msgs else None
    return {
        **{k: v for k, v in chat.items() if k != "messages"},
        "messages": [],
        "message_count": len(msgs),
        "preview": (str(last.get("content") or "")[:200] if last else ""),
    }


def _row_to_chat(row) -> Optional[Dict[str, Any]]:
    doc = db.loads(row["doc"])
    return doc if isinstance(doc, dict) else None


# ── reads ────────────────────────────────────────────────────────────────────

def list_chats(workspace: Optional[str] = None, limit: int = 200,
               offset: int = 0) -> Dict[str, Any]:
    """A page of conversations, newest activity first, without transcripts."""
    where, params = "", []
    if workspace:
        where = "WHERE workspace = ?"
        params.append(workspace)
    conn = db.get_conn()
    total = conn.execute(f"SELECT COUNT(*) FROM chats {where}", params).fetchone()[0]
    rows = conn.execute(
        f"SELECT doc FROM chats {where} "
        "ORDER BY COALESCE(updated_at, created_at) DESC, rowid DESC LIMIT ? OFFSET ?",
        [*params, max(1, min(int(limit), 500)), max(0, int(offset))],
    ).fetchall()
    items = [_summary(c) for c in (_row_to_chat(r) for r in rows) if c is not None]
    return {"items": items, "total": total, "limit": limit, "offset": offset}


def get_chat(chat_id: str) -> Optional[Dict[str, Any]]:
    """One conversation with its full transcript."""
    row = db.get_conn().execute(
        "SELECT doc FROM chats WHERE chat_id = ?", (str(chat_id),)).fetchone()
    return _row_to_chat(row) if row else None


def chat_ids() -> List[str]:
    rows = db.get_conn().execute("SELECT chat_id FROM chats").fetchall()
    return [r["chat_id"] for r in rows]


# ── writes ───────────────────────────────────────────────────────────────────

def save_chat(chat: Dict[str, Any]) -> Dict[str, Any]:
    """Create or replace one conversation. Returns what was stored.

    ``created_at`` survives a rewrite: the client sends the conversation it
    holds, and a reload that lost the original timestamp must not reorder the
    history.
    """
    chat_id = str(chat.get("id") or "").strip()
    if not chat_id:
        raise ValueError("chat id is required")

    with db.transaction() as conn:
        row = conn.execute("SELECT doc FROM chats WHERE chat_id = ?",
                           (chat_id,)).fetchone()
        existing = _row_to_chat(row) if row else None
        doc = _fit({
            **chat,
            "id": chat_id,
            "created_at": (chat.get("created_at")
                           or (existing or {}).get("created_at") or _now()),
            "updated_at": _now(),
        })
        conn.execute(
            "INSERT OR REPLACE INTO chats "
            "(chat_id, title, workspace, project_id, agent_id, flow_id, team_id,"
            " target_mode, origin, message_count, created_at, updated_at, doc) "
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            (chat_id, *[doc.get(c) for c in _COLUMNS], len(_messages(doc)),
             doc.get("created_at"), doc.get("updated_at"), db.dumps(doc)),
        )
    return doc


def append_turn(chat_id: str, *, run_id: str, user_message: str,
                agent_message: str, agent_id: Optional[str] = None,
                usage: Optional[Dict[str, Any]] = None,
                duration_ms: Optional[int] = None) -> bool:
    """Add one completed exchange to a stored chat. Returns whether it was added.

    This is the safety net under the browser, not the normal path: the page that
    ran the turn writes its own, richer version of these bubbles (tool traces,
    touched files, rendered views). It matters when that page is gone — a tab
    closed mid-answer used to take the answer with it, although the run had
    completed and been recorded.

    Two guards keep it from fighting the browser. A chat that does not exist is
    left alone, because a conversation the dashboard never created (a Telegram
    thread, an instance delivery) should not appear in the sidebar. A run whose
    id is already on a message is skipped, so whichever writer got there first
    holds the turn and the other does not duplicate it.
    """
    if not chat_id or not run_id:
        return False
    with db.transaction() as conn:
        row = conn.execute("SELECT doc FROM chats WHERE chat_id = ?",
                           (str(chat_id),)).fetchone()
        chat = _row_to_chat(row) if row else None
        if chat is None:
            return False
        messages = _messages(chat)
        if any(str(m.get("run_id") or "") == str(run_id) for m in messages):
            return False
        turn = []
        if (user_message or "").strip():
            turn.append({"id": f"srv-u-{run_id}", "role": "user",
                         "content": user_message})
        turn.append({
            "id": f"srv-a-{run_id}", "role": "agent", "agent_id": agent_id,
            "content": agent_message, "error": False, "run_id": str(run_id),
            "inbound_tokens": (usage or {}).get("inbound_tokens"),
            "outbound_tokens": (usage or {}).get("outbound_tokens"),
            "total_tokens": (usage or {}).get("total_tokens"),
            "duration_ms": duration_ms,
        })
        doc = _fit({**chat, "messages": messages + turn, "updated_at": _now()})
        conn.execute(
            "UPDATE chats SET message_count = ?, updated_at = ?, doc = ? WHERE chat_id = ?",
            (len(_messages(doc)), doc["updated_at"], db.dumps(doc), str(chat_id)),
        )
    return True


def delete_chat(chat_id: str) -> bool:
    with db.transaction() as conn:
        cur = conn.execute("DELETE FROM chats WHERE chat_id = ?", (str(chat_id),))
    return cur.rowcount > 0


def import_chats(items: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Take in conversations a browser still holds, without clobbering.

    Used once per browser profile to lift the old ``localStorage`` history into
    the database. An id already in the store is left alone: the stored copy is
    the one other devices have been writing to, and the browser's copy may be a
    stale snapshot of the same conversation.
    """
    known = set(chat_ids())
    imported, skipped = 0, 0
    for item in items or []:
        if not isinstance(item, dict) or not str(item.get("id") or "").strip():
            skipped += 1
            continue
        if str(item["id"]) in known:
            skipped += 1
            continue
        save_chat(item)
        known.add(str(item["id"]))
        imported += 1
    return {"imported": imported, "skipped": skipped}


__all__ = ["list_chats", "get_chat", "save_chat", "delete_chat", "append_turn",
           "import_chats",
           "chat_ids", "MAX_MESSAGES", "MAX_DOC_BYTES"]

File: dashboard/frontend/src/components/chatStore.js
import { useCallback, useEffect, useRef, useState } from 'react';
import { deleteChat, getChat, importChats, listChats, saveChat } from '../api';
import { useChannel, useStream } from './stream';

/**
 * The Chat page's conversations, stored on the server.
 *
 * They used to live in `localStorage`, which made the main record this service
 * produces the one thing it did not keep: tied to a single browser profile,
 * gone with the site data, and — once the ~5 MB quota was reached — trimmed
 * oldest-first with nothing said. They are rows in the hub's database now
 * (`/api/chats`), beside the runs they spawned.
 *
 * The component's state shape is untouched: the same `conversations` array and
 * the same `setConversations` the whole page already mutates in a hundred
 * places. Only the persistence underneath it changed, which is why this is a
 * hook and not a rewrite of the page.
 *
 * Three things are worth knowing if this is ever changed:
 *
 * * **The list arrives without transcripts.** A chat's messages are fetched
 *   when that chat is opened. Loading every bubble of every conversation to
 *   draw a sidebar would be the old problem in a new place.
 * * **Writes are debounced per conversation.** A streaming turn rewrites its
 *   bubble on every token; saving each frame would be a PUT per token. Changes
 *   settle for {@link SYNC_DEBOUNCE_MS}, and a long stream is still flushed
 *   every {@link SYNC_MAX_WAIT_MS} so leaving mid-answer does not lose it.
 * * **An unhydrated conversation is never written back.** It holds an empty
 *   `messages` array that is absence of knowledge, not an empty transcript;
 *   saving it would erase the stored one.
 * * **Someone else's save is picked up.** A write announces itself on the
 *   conversation's channel, so the same chat open in another tab or on another
 *   device reloads instead of drifting. A reload while this tab has unsaved
 *   changes of its own would throw them away, so it waits.
 */

/** The browser store this replaces. Read once, then imported server-side. */
export const LEGACY_STORAGE_KEY = 'agent_hub_chats_v1';
/** Set after the legacy payload has been handed to the server. */
export const MIGRATED_FLAG_KEY = 'agent_hub_chats_migrated_v1';

/** How long a conversation must sit unchanged before it is written. */
export const SYNC_DEBOUNCE_MS = 1000;
/** …and the longest a continuously changing one may go unwritten. */
export const SYNC_MAX_WAIT_MS = 8000;
/** How often a reload deferred by a busy tab checks whether it can happen. */
const DEFERRED_RELOAD_POLL_MS = 500;
/** Consecutive failed writes before the hook stops retrying and reports it. */
const MAX_SYNC_ATTEMPTS = 3;
/** How long it then waits before giving the backend another chance. */
const RETRY_COOLDOWN_MS = 30000;

// Per-message fields that are large and only meaningful while the turn is live
// (the Build-view timeline, the running-tool indicator, the reasoning ticker).
// They accumulate every streamed token and every tool input/output, and they are
// reconstructable from the run's server-side log, so they are not stored.
export const TRANSIENT_MSG_FIELDS = ['timeline', 'running_tool', 'thinking_live'];

/** The conversation as it is stored: no transient bubble state, no local flags. */
export function stripForStorage(conv) {
  const { _hydrated, message_count: _count, preview: _preview, ...rest } = conv;
  return {
    ...rest,
    messages: (conv.messages || []).map((m) => {
      const copy = { ...m };
      for (const f of TRANSIENT_MSG_FIELDS) delete copy[f];
      return copy;
    }),
  };
}

const serialize = (conv) => JSON.stringify(stripForStorage(conv));

function readLegacy() {
  try {
    return JSON.parse(localStorage.getItem(LEGACY_STORAGE_KEY) || '[]') || [];
  } catch {
    return [];
  }
}

/**
 * Hand the browser's old history to the server, once per browser profile.
 *
 * The raw payload is left in `localStorage` on purpose: the import is additive
 * (ids already stored are skipped), so keeping the old copy costs nothing and
 * is the only safety net if this ever imports the wrong thing.
 */
async function migrateLegacyChats() {
  let alreadyDone = false;
  try { alreadyDone = localStorage.getItem(MIGRATED_FLAG_KEY) === '1'; } catch { return; }
  if (alreadyDone) return;
  const legacy = readLegacy();
  if (!legacy.length) {
    try { localStorage.setItem(MIGRATED_FLAG_KEY, '1'); } catch { /* storage unavailable */ }
    return;
  }
  // A failure here is left unmarked so the next load tries again — the history
  // is the user's, and dropping it because the backend blinked is not an option.
  await importChats(legacy);
  try { localStorage.setItem(MIGRATED_FLAG_KEY, '1'); } catch { /* storage unavailable */ }
}

/**
 * @param {string|null} currentConvId the conversation being read, which is the
 *   one whose transcript is fetched.
 * @returns {{conversations: Array, setConversations: Function, ready: boolean,
 *            removeConversation: Function, syncError: boolean, flush: Function}}
 */
export function useConversationStore(currentConvId, { paused = false } = {}) {
  const { clientId } = useStream();
  const [conversations, setConversations] = useState([]);
  const [ready, setReady] = useState(false);
  const [syncError, setSyncError] = useState(false);

  // Latest state, readable from the debounce timer without re-subscribing it.
  const convsRef = useRef(conversations);
  // id → the document last written, so an unchanged conversation is not resent.
  const savedRef = useRef(new Map());
  // id → the object last looked at. A streaming turn replaces its conversation
  // object on every token, so the change check is an identity comparison here
  // and a serialization only when a write is actually about to happen.
  const seenRef = useRef(new Map());
  const dirtyRef = useRef(new Set());
  const timerRef = useRef(null);
  const deadlineRef = useRef(0);
  const attemptsRef = useRef(0);
  // When the attempt budget is spent, when it may be tried again.
  const blockedUntilRef = useRef(0);
  const hydratingRef = useRef(new Set());
  const clientIdRef = useRef(clientId);
  // A save that landed while this tab was busy, replayed once it is free.
  const pendingReloadRef = useRef(null);
  const reloadTimerRef = useRef(null);
  const pausedRef = useRef(paused);

  useEffect(() => { convsRef.current = conversations; }, [conversations]);
  useEffect(() => { clientIdRef.current = clientId; }, [clientId]);
  useEffect(() => { pausedRef.current = paused; }, [paused]);

  // ---- initial load ----
  useEffect(() => {
    let cancelled = false;
    (async () => {
      try {
        await migrateLegacyChats();
      } catch { /* retried on the next load */ }
      let items = null;
      try {
        const { data } = await listChats({ limit: 200 });
        items = data?.items || [];
      } catch {
        // Backend unreachable: fall back to whatever this browser still holds,
        // so an offline tab shows a history instead of an empty page. Nothing
        // is written back from that state — the records stay unhydrated.
        items = readLegacy();
      }
      if (cancelled) return;
      setConversations(items.map((c) => ({ ...c, messages: c.messages || [], _hydrated: false })));
      setReady(true);
    })();
    return () => { cancelled = true; };
  }, []);

  // ---- fetch the open conversation's transcript ----
  useEffect(() => {
    if (!ready || !currentConvId) return undefined;
    const conv = conversations.find((c) => c.id === currentConvId);
    if (!conv || conv._hydrated !== false) return undefined;
    // A Telegram thread's transcript is not this store's to serve: the page
    // rebuilds it from the runs recorded against the binding, workspace slice
    // and all. Leaving it unhydrated keeps that rebuild from being written back
    // over itself and keeps the two copies from racing on open.
    if (conv.origin === 'telegram') return undefined;
    const hydrating = hydratingRef.current;
    if (hydrating.has(currentConvId)) return undefined;
    hydrating.add(currentConvId);
    let cancelled = false;
    (async () => {
      let loaded = null;
      try {
        const { data } = await getChat(currentConvId);
        loaded = data || {};
      } catch (e) {
        // A chat the server has never heard of (a record this browser still
        // holds from before the store existed) is as hydrated as it will get.
        // Any other failure means the transcript is unknown, not empty — the
        // record stays unhydrated so the empty `messages` can never be written
        // back over the stored one.
        if (e?.response?.status !== 404) { hydrating.delete(currentConvId); return; }
        loaded = {};
      }
      if (cancelled) return;
      setConversations((prev) => prev.map((c) => {
        if (c.id !== currentConvId) return c;
        const merged = { ...c, ...loaded, messages: loaded.messages || c.messages || [], _hydrated: true };
        // Record what was read as already written and already seen, so opening
        // a chat neither sends it straight back nor counts as an edit that has
        // to settle before anything else can happen.
        savedRef.current.set(currentConvId, serialize(merged));
        seenRef.current.set(currentConvId, merged);
        return merged;
      }));
      hydrating.delete(currentConvId);
    })();
    return () => { cancelled = true; hydrating.delete(currentConvId); };
  }, [ready, currentConvId, conversations]);

  // ---- someone else saved this chat ----
  const reloadChat = useCallback(async (chatId) => {
    if (!chatId) return;
    let loaded = null;
    try {
      const { data } = await getChat(chatId);
      loaded = data;
    } catch {
      return;  // keep what this tab has rather than emptying it on a failed read
    }
    if (!loaded) return;
    setConversations((prev) => prev.map((c) => {
      if (c.id !== chatId) return c;
      const merged = { ...c, ...loaded, messages: loaded.messages || [], _hydrated: true };
      // What was just read is, by definition, what is stored: record it as both
      // seen and saved so taking it in is not mistaken for a local change and
      // written straight back.
      savedRef.current.set(chatId, serialize(merged));
      seenRef.current.set(chatId, merged);
      return merged;
    }));
  }, []);

  const busy = useCallback((chatId) => pausedRef.current || dirtyRef.current.has(chatId), []);

  // A reload while this tab has a turn running, or changes not yet written,
  // would throw them away — so it waits and is retried until the tab is free.
  const deferReload = useCallback((chatId) => {
    pendingReloadRef.current = chatId;
    if (reloadTimerRef.current) return;
    reloadTimerRef.current = setInterval(() => {
      const id = pendingReloadRef.current;
      if (id && busy(id)) return;
      clearInterval(reloadTimerRef.current);
      reloadTimerRef.current = null;
      pendingReloadRef.current = null;
      if (id) reloadChat(id);
    }, DEFERRED_RELOAD_POLL_MS);
  }, [busy, reloadChat]);

  useChannel(currentConvId ? `chat:${currentConvId}` : null, (event) => {
    if (event?.type !== 'chat_saved') return;
    if (event.origin_client && event.origin_client === clientIdRef.current) return;
    if (busy(currentConvId)) { deferReload(currentConvId); return; }
    reloadChat(currentConvId);
  });

  useEffect(() => () => {
    if (reloadTimerRef.current) clearInterval(reloadTimerRef.current);
  }, []);

  // ---- write the changed conversations ----
  // `flush` schedules its own retry, so the timer goes through a ref rather
  // than through the callback itself, which cannot reference itself.
  const flushRef = useRef(() => {});
  const schedule = useCallback((wait) => {
    if (timerRef.current) clearTimeout(timerRef.current);
    timerRef.current = setTimeout(() => { timerRef.current = null; flushRef.current(); }, wait);
  }, []);

  const flush = useCallback(() => {
    if (timerRef.current) { clearTimeout(timerRef.current); timerRef.current = null; }
    deadlineRef.current = 0;
    const ids = [...dirtyRef.current];
    dirtyRef.current.clear();
    for (const id of ids) {
      const conv = convsRef.current.find((c) => c.id === id);
      if (!conv || conv._hydrated === false) continue;  // deleted, or never loaded
      const body = serialize(conv);
      // Changed object, identical document: a re-render, or the transcript that
      // was just read back. Nothing to send.
      if (savedRef.current.get(id) === body) continue;
      savedRef.current.set(id, body);
      saveChat(JSON.parse(body), clientIdRef.current)
        .then(() => { attemptsRef.current = 0; setSyncError(false); })
        .catch(() => {
          // Forget the snapshot so the change is seen again, and retry until
          // the attempt budget runs out — a backend that is down should not
          // turn into a write every second for the rest of the session.
          savedRef.current.delete(id);
          attemptsRef.current += 1;
          if (attemptsRef.current >= MAX_SYNC_ATTEMPTS) {
            blockedUntilRef.current = Date.now() + RETRY_COOLDOWN_MS;
            setSyncError(true);
            return;
          }
          dirtyRef.current.add(id);
          if (!timerRef.current) schedule(SYNC_DEBOUNCE_MS);
        });
    }
  }, [schedule]);

  useEffect(() => { flushRef.current = flush; }, [flush]);

  useEffect(() => {
    if (!ready) return;
    let dirty = false;
    for (const conv of conversations) {
      if (!conv?.id || conv._hydrated === false) continue;
      if (seenRef.current.get(conv.id) === conv) continue;
      seenRef.current.set(conv.id, conv);
      dirtyRef.current.add(conv.id);
      dirty = true;
    }
    if (!dirty) return;
    if (attemptsRef.current >= MAX_SYNC_ATTEMPTS) {
      // The budget is spent. A backend that comes back should be picked up, so
      // the next change after the cooldown starts a fresh round rather than the
      // tab staying read-only until it is reloaded.
      if (Date.now() < blockedUntilRef.current) return;
      attemptsRef.current = 0;
    }
    // A conversation being streamed into changes on every token, so the debounce
    // would keep sliding; the deadline is what guarantees it is written anyway.
    if (!deadlineRef.current) deadlineRef.current = Date.now() + SYNC_MAX_WAIT_MS;
    schedule(Math.max(0, Math.min(SYNC_DEBOUNCE_MS, deadlineRef.current - Date.now())));
  }, [conversations, ready, schedule]);

  // Leaving the page is the one moment a pending write must not wait out its
  // debounce. `visibilitychange` fires where `beforeunload` is unreliable
  // (mobile, tab discard), and unmount covers navigating within the app.
  useEffect(() => {
    const onHide = () => { if (document.visibilityState === 'hidden') flushRef.current(); };
    const onPageHide = () => flushRef.current();
    document.addEventListener('visibilitychange', onHide);
    window.addEventListener('pagehide', onPageHide);
    return () => {
      document.removeEventListener('visibilitychange', onHide);
      window.removeEventListener('pagehide', onPageHide);
      flushRef.current();
    };
  }, []);

  // ---- delete ----
  const removeConversation = useCallback((id) => {
    dirtyRef.current.delete(id);
    savedRef.current.delete(id);
    seenRef.current.delete(id);
    setConversations((prev) => prev.filter((c) => c.id !== id));
    // The row is the record; a failed delete leaves it, and the next load shows
    // it again rather than pretending it is gone.
    return deleteChat(id).catch(() => {});
  }, []);

  return {
    conversations, setConversations, ready, removeConversation, syncError, flush,
    reloadChat,
  };
}

export default useConversationStore;
